File: cpp-oo/src/Moon.cpp
#include "Moon.h"
#include <stdexcept>
#include <iostream>
#include <math.h>

namespace moon {

    const double Moon::MOON_ORBIT_TIME=27.322;

    Moon::Moon()
      : Moon(239000.0, DEFAULT_ORBITS, 1079.4,"silver",0)
      {
      }

    Moon::Moon(double distance, const std::string &orbits,
                int size, const std::string &color, int phase) 
        : Satellite(distance,orbits), m_size(size), m_color(color), m_phase(phase),
        m_time((MOON_ORBIT_TIME/4.0)*(phase+0.5))
    {
          std::cout << "Moon@" << (void*) this << " constructed." << std::endl;
    }

    Moon::~Moon() {
          std::cout << "Moon@" << (void*) this << " destructed." << std::endl;
    }
    bool Moon::isNewMoon() const {
        return m_phase == 0;
    }
    void Moon::cycle() {
        m_phase = (m_phase + 1) % 4;
        m_time += MOON_ORBIT_TIME/4.0;
    }

    void Moon::advance(double amount) {
        m_time += amount;
        m_phase = int(4*(m_time/MOON_ORBIT_TIME)) % 4;
    }

    int Moon::phase() const {
        return m_phase;
    }
    const std::string& Moon::color() const {
        return m_color;
    }
    void Moon::color(std::string value) {
        if (!(value == "silver" || value == "white" )) {
            throw std::invalid_argument("invalide color");
        }
        m_color = value;
    }
    
    double Moon::time() const { return m_time; }

    int Moon::size() const {
        return m_size;
    }

}

File: docs/moon.mjs
import { SatelliteModel } from './satellite.mjs'



export class MoonModel extends SatelliteModel {
    constructor(config) {
        super(config.distance, config.orbits);
        this._size = config.size;
        this._color = config.color;
        this._phase = config.phase;
        this._observers = new Array();
    }

    addObserver(observer) {
        this._observers.push(observer);
    }

    _observe() {
        this._observers.forEach((observer) => observer.onMoonModelChange(this));
    }


    get phase() {
        return this._phase
    }

    set phase(value) {
        if (this._phase != value) {
            this._phase = value;
            this._observe();
        }
    }

    get color() {
        return this._color
    }

    set color(value) {
        if (!["white", "silver"].includes(value)) {
            throw Error('invalid value')
        }
        if (this._color != value) {
            this._color = value;
            this._observe();
        }
    }

    get size() {
        return this._size
    }
}

export class MoonController {
    constructor(model) {
        this._model = model;
    }
    get model() {
        return this._model;
    }

    cycle() {
        this._model.phase = (this._model.phase + 1) % 4;
    }

    isNewMoon() {
        return this._model.phase == 0
    }
}

export class MoonView {
    createCanvas() {
        const canvas = document.createElement("canvas");
        canvas.setAttribute("id", this._id + "-canvas");
        canvas.width = 200;
        canvas.height = 200;
        canvas.setAttribute("style", "border: 1px solid black");
        return canvas;
    }

    constructor(model, controller, div, id, config) {
        this._model = model;
        this._controller = controller;
        this._div = div;
        this._id = id;
        this._div.innerHTML = "Moon(id=" + this._id + ", color=" + this._color + ")";
        this._canvas = this.createCanvas();
        this._div.appendChild(this._canvas);
        const me = this;
        this._model.addObserver(this);
        setInterval(() => { me.cycle(); }, 1000);
    }

    onMoonModelChange() {
        this.redraw();
    }

    redraw() {
        const canvas = this._canvas;
        canvas.width = canvas.width;
        const context = this._canvas.getContext("2d");
        const x = canvas.width / 2;
        const y = canvas.height / 2;
        const r = 0.80*Math.min(x,y);
        let sAngle = 0;
        let eAngle = Math.PI*2;
        const counterclockwise = false;

        if (this._phase == 1) {
            eAngle = Math.PI;
        } else if (this._phase == 3) {
            sAngle = Math.PI;
        }

        sAngle += Math.PI/2;
        eAngle += Math.PI/2;

        context.beginPath();
        context.arc(x,y,r,sAngle,eAngle,counterclockwise);
        context.fillStyle=(this._phase == 0) ? "black" : this._color; 
        context.fill();
    }

}

class Moon

export function MoonInstantiate() {
    const divs = document.getElementsByClassName("moon");
    for (let i = 0; i < divs.length; ++i) {
        const div = divs[i];
        const id = "moon-" + i;
        const config = JSON.parse(div.getAttribute("data-config"));
        const moon = new Moon(div, id, config)
    }
}
